src/v/bytes/iobuf.h

```diff
@@ -305,7 +373,18 @@ inline void iobuf::reserve_memory(size_t reservation) {
     }
     oncore_debug_verify(_verify_shard);
     const size_t last_asz = last_allocation_size();
-    if (b.size() <= available_bytes() || b.size() <= last_asz) {
+    // The following is a heuristic to decide between copying and zero-copy
+    // append of the source buffer. The rule we apply is if the buffer we are
+    // appending is of the same size or smaller than last allocation we use
+    // a copy. This effecitvely linearizes a series of smaller buffers when
+    // they are appended to this one. However, if the incoming buffer is
+    // already at (or above) the maximum fragment size, we use zero copy as
+    // copying it provides almost no linearization benefits and appending
+    // full-sized fragments is in practice a common operation when buffers
+    // grow beyond the maximum fragment size.
+    if (
+      b.size() <= last_asz
```

Comment:
So you kept `b.size() <= last_asz` instead of `b.size() <= available_bytes()` to still force linearization even if it does require an extra alloc?

Reply:
I think `b.size() <= available_bytes()` is subsumed by `b.size() <= last_asz`, i.e., when former is true the latter is also true (last_asz is always >= available_bytes()), so it can simply be deleted. I.e., this deletion was just a performance optimization.

Not sure if that addresses your question?

Reply:
Another answer is that using only `b.size() < available_bytes()` would be a very large change in behavior and break a key property: that when you append a series of buffers, the growth strategy of the io_buf is still doubling (until the 128K cap).

If the check was just `b.size() < available_bytes()` then appending N 1 byte buffer when `avaialble_size()` was initially zero would actually result in N 1-byte fragments. `io_buf` is trying to avoid that (and there are tests that this doesn't happen).

So we want to preserve the "doubling allocations" behavior sort of regardless of the sequence of appended buffers, which means more copying/linearization until (after this change) we hit the cap.

Reply:
Yes I meant rather only keeping the `b.size() <= available_bytes()` check to avoid linearization where it would require extra allocations.

However that would be a bigger functional change so I think keeping it as is is fine but was just wondering whether you had any further thoughts on it.

Reply:
I _did_ have further thoughts on it: I think a slightly more aggressive version of this change would use zero-copy in more cases, e.g., maybe if the buffer is more than half the size of the max alloc (128K) or some other threshold if it is somehow "similar" in size to the last allocation (e.g., maybe if it's more than half that size). With the latter type of change though you have to avoid it "going linear" which is what happens if you just implement that (or `< last_asz` instead of `<= last_asz` as I tried originally) because that results in no more doubling.

It seems like the two most important properties are:

1) Keep the doubling growth
2) Do zero copy "at the limit" of large iobufs

This fixes (2) without impacting (1) and beyond that it's just tuning (though I think we should do the tuning which is why I did not close #15769.

Reply:
Even this change is possibly worse on some dimension than the existing behaviors for _some_ patterns, e.g., appending allocating 64K+1, alternating with 128K. The old behavior would linearize it to full 128K fragments, involving copying every byte but being fully efficient in memory _under the "unique owner" assumption_ (i.e., the source buffers all go away).

The new behavior will result in (for each 2 appends) 128K capacity fragments, but with every other one being half-empty.

Similarly with 1-byte and 128K alternating appends, the new one will be mostly memory efficient but with 1 byte buffers alternating with 128K ones (and needing an extra free+alloc to trim down to 1 byte).

I don't think you can totally solve these problems without an oracle though: an oracle that knows _both_ about the future append pattern and also the lifetime of the source buffers (i.e., zero-copy is relatively more favorable if the source buffer lifetime extends beyond the lifetime of this buffer, since then it really saves you memory footprint).

Even w/o that though some of the methods could be smarter: like this `append(io_buf)` method just has a loop where it appends the underlying fragments 1-by-1 and then within that append (which hits the code in this PR) we have to make a local decision about copying or not, and growth: however the outer append method already knows the full size of the incoming `iobuf` so it could just see that it's 400K or whatever and skip straight to 128K fragments (assuming linearizatoin was what you wanted), avoiding a bunch of pointless small allocations.